vslib/src/SwitchStateBase.cpp

```diff
@@ -2424,3 +2439,176 @@ void SwitchStateBase::debugSetStats(
         m_countersMap[key][kvp.first] = kvp.second;
     }
 }
+
+sai_status_t SwitchStateBase::initialize_voq_switch_objects(
```

Comment:
seems like this is main initialization function for voq, but it's not executed anywhere in SwitchStateBae (if it's generic for both Brcm/Mlnx switches)
should be called in: SwitchStateBase::initialize_default_objects()

Reply:
you also want to take a look at warm_boot_initialize_objects

Reply:
For comment 1: The VOQ switch objects initialized here are not default objects. They need specific configuration that should be supplied by orchagent during switch creation for all types of VoQ switches. The configurations are in the attribute list. Right now this is called from VirtualSwitchSaiInterface class from the place where init_swith() is called. This is the bottom most place where attribute list is visible. This is the reason why this API is invoked from here. If it is still preferred to get this API invoked from SwitchStateBase form initialize_default_object(), I can modify the init_switch() and initialize_default_objects() carry forward the attr_count and attr_list.

For comment 2: I had a look at the warm_boot_initialize_objects(). There is refreshing of m_port_list(). An equivalent for VoQ system will be refreshing m_system_port_list. As of now, there is nothing in the SAI or hardware or SDK that changes the system port info once the system is initialized. It is required that all system ports of the VoQ system must be pre-configured during bootup, irrespective of whether cold boot or warmboot. So at least for now, no need to do anything for warm boot as far as VOQ switch objects are concerned.

Reply:
Ad 1: If you need some extra action to be performed when creating switch, move this initialization to special case in function: sai_status_t SwitchStateBase::create, there are already some custom functions like port or hostif  that require some extra steps, and that would be the way to go, although, since switch object is special,  there could be some other issues with this, but please try to make it like port/hostif

Ad 2: ok

Reply:
so far initialize_default_objects did not need attr_count/attr_list since creation of internal objects didnt depend on any attributes, in virtual switch each attribute is just stored in map, so there are no actual actions needed. in your case this is different, since you create objects system ports inside switch create based on switch attributes.  This maybe require to pass attr list and count to init_switch and pass to constructiors of SwitchStateBase
is system port object related anyhow to SAI_OBJECT_TYPE_PORT ?

Reply:
Yes. There is system port config representing every physical port in the system.

Reply:
> **there can be issue here, since initialize_default_objects is called inside VirtualSwitchSaiInterface::init_switch and this is called inside VirtualSwitchSaiInterface::create, which after init_switch calls ss->create(..) which creates switch, and this actually works, since create() will just populate internal string values of the switch, and it can't call initialize_voq inside this create function, let leave this for now and let's use that initialize after init_switch() as you had so far**

o.k. I'll leave it like how it is now. I mention this as response to comment in PR #702. Thanks.